File: engine/chessboard.cpp
#include "chessboard.h"
#include "bitboard.h"
#include "move.h"
#include "move_generator.h"
#include "random.h"
#include <algorithm>
#include <sstream>
#include <string>

uint64_t PieceKeys[2][6][64] = {0};
uint64_t CastleKeys[4] = {0};
uint64_t EnPassantKeys[8] = {0};
uint64_t whiteToMoveKey = 0;

// generate key for position hash
void init_keys() {
    for (int i = 0; i < 8; i++) {
        EnPassantKeys[i] = randInt64();
    }
    for (int i = 0; i < 4; i++) {
        CastleKeys[i] = randInt64();
    }
    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < 6; j++) {
            for (int k = 0; k < 64; k++) {
                PieceKeys[i][j][k] = randInt64();
            }
        }
    }
    whiteToMoveKey = randInt64();
}

std::string ChessBoard::to_string() const {
    std::string board;

    for (int i = 0; i < 64; i++) {
        if (pawns_.get(i)) {
            board += 'p';
        } else if (knights_.get(i)) {
            board += 'n';
        } else if (bishops_.get(i)) {
            board += 'b';
        } else if (rooks_.get(i)) {
            board += 'r';
        } else if (queens_.get(i)) {
            board += 'q';
        } else if (kings_.get(i)) {
            board += 'k';
        } else {
            board += '.';
        }
        if (white_pieces_.get(i)) {
            // convert last char to uppercase
            board.back() -= 32;
        }
        if (i % 8 == 7) {
            board += '\n';
        } else {
            board += ' ';
        }
    }

    std::ostringstream result;
    for (int i = 7; i >= 0; i--) {
        result << i + 1 << "  ";
        for (int j = 0; j < 16; j++) {
            result << board[i * 16 + j];
        }
    }

    result << "   a b c d e f g h\n\n";

    return result.str();
}

bool ChessBoard::act(Move move, bool update) {
    Square from = move.from_;
    Square to = move.to_;
    char promotion = move.promotion_;

    update_draw_condition(from, to);
    check_en_passant(from, to);
    check_promotion(promotion, from);
    castling(from, to);
    pawns_.update(from, to);
    knights_.update(from, to);
    bishops_.update(from, to);
    rooks_.update(from, to);
    queens_.update(from, to);
    kings_.update(from, to);
    white_pieces_.update(from, to);
    black_pieces_.update(from, to);
    all_pieces_.update(from, to);

    position_hash_history_.push_back(generate_hash());
    player_ = (player_ == Player::White) ? Player::Black : Player::White;

    if (update) {
        update_game_state();
        std::cout << to_string() << std::endl;
    }
    fullmove_number_++;

    return true;
}

void ChessBoard::check_en_passant(Square from, Square to) {
    // check if en passant is played
    if (pawns_.get(from) && en_passant_.get(to)) {
        // clear captured pawn
        if (white_pieces_.get(from)) {
            pawns_.clear(to.square_ - 8);
            black_pieces_.clear(to.square_ - 8);
            all_pieces_.clear(to.square_ - 8);
        } else if (black_pieces_.get(from)) {
            pawns_.clear(to.square_ + 8);
            white_pieces_.clear(to.square_ + 8);
            all_pieces_.clear(to.square_ + 8);
        }
    }
    en_passant_.reset();
    // check for double pawn moves
    if (pawns_.get(from) && (abs(from.rank_ - to.rank_) == 2)) {
        en_passant_.set((from.square_ + to.square_) / 2);
    }
}

void ChessBoard::check_promotion(char promotion, Square from) {
    if (promotion) {
        switch (promotion) {
        case 'q':
            queens_.set(from);
            break;
        case 'r':
            rooks_.set(from);
            break;
        case 'b':
            bishops_.set(from);
            break;
        case 'n':
            knights_.set(from);
            break;
        default:
            std::cerr << "invalid promotion piece" << std::endl;
        }
        pawns_.clear(from);
    }
}

Bitboard ChessBoard::generate_moves(Square from) const {
    Bitboard our_pieces =
        white_pieces_.get(from) ? white_pieces_ : black_pieces_;
    Bitboard moves(0);

    if (pawns_.get(from)) {
        if (white_pieces_.get(from)) {
            moves = generate_white_pawn_moves(from, all_pieces_,
                                              black_pieces_ | en_passant_);
        } else if (black_pieces_.get(from)) {
            moves = generate_black_pawn_moves(from, all_pieces_,
                                              white_pieces_ | en_passant_);
        }
    } else if (knights_.get(from)) {
        moves = generate_knight_moves(from);
    } else if (bishops_.get(from)) {
        moves = generate_bishop_moves(from, all_pieces_);
    } else if (rooks_.get(from)) {
        moves = generate_rook_moves(from, all_pieces_);
    } else if (queens_.get(from)) {
        moves = generate_queen_moves(from, all_pieces_);
    } else if (kings_.get(from)) {
        moves = generate_king_moves(from);
    } else {
        moves = Bitboard(0);
    }

    // Remove moves that would capture our own pieces
    moves &= ~our_pieces;

    return moves;
}

Bitboard ChessBoard::generate_legal_moves(Square from) const {
    Bitboard moves = generate_moves(from);
    // add castling moves
    if (kings_.get(from)) {
        if (from == 4) {
            Bitboard all_black_moves = Bitboard(0);
            for (auto from : black_pieces_) {
                all_black_moves |= generate_moves(from);
                if (pawns_.get(from)) {
                    all_black_moves.set(from.square_ + 7);
                    all_black_moves.set(from.square_ + 9);
                }
            }

            // can castle if no squares occupied or attacked between king and
            // rook, and castling rights are set
            bool can_white_castle_kingside =
                (all_black_moves & white_kingside_squares).empty() &&
                (all_pieces_ & (white_kingside_squares & ~kings_)).empty() &&
                (castling_rights_ & 1);
            bool can_white_castle_queenside =
                (all_black_moves & white_queenside_squares).empty() &&
                (all_pieces_ & (white_queenside_squares & ~kings_)).empty() &&
                (castling_rights_ & 2);

            if (can_white_castle_kingside) {
                moves.set(6);
            }
            if (can_white_castle_queenside) {
                moves.set(2);
            }
        } else if (from == 60) {
            Bitboard all_white_moves = Bitboard(0);
            for (auto from : white_pieces_) {
                all_white_moves |= generate_moves(from);
                // pawn attacks on castling squares even if no pieces are there
                if (pawns_.get(from)) {
                    all_white_moves.set(from.square_ + 7);
                    all_white_moves.set(from.square_ + 9);
                }
            }

            // can castle if no squares occupied or attacked between king and
            // rook, and castling rights are set
            bool can_black_castle_kingside =
                (all_white_moves & black_kingside_squares).empty() &&
                (all_pieces_ & (black_kingside_squares & ~kings_)).empty() &&
                (castling_rights_ & 4);

            bool can_black_castle_queenside =
                (all_white_moves & black_queenside_squares).empty() &&
                (all_pieces_ & (black_queenside_squares & ~kings_)).empty() &&
                (castling_rights_ & 8);

            if (can_black_castle_kingside) {
                moves.set(62);
            }
            if (can_black_castle_queenside) {
                moves.set(58);
            }
        }
    }
    // remove moves that would put our king in check
    // e.g. pins, illegal king moves
    for (auto to : moves) {
        ChessBoard temp_board = *this;
        temp_board.act(Move(from, to, '\0'), false);
        if (temp_board.is_player_in_check(player_)) {
            moves.clear(to);
        }
    }
    return moves;
}

bool ChessBoard::is_player_in_check(Player player) const {
    Bitboard their_pieces = this->their_pieces(player);
    Bitboard our_king = our_pieces(player) & kings_;
    Square our_king_position = Square(our_king.getLSB());

    for (auto from : their_pieces) {
        Bitboard moves = generate_moves(from);
        if (moves.get(our_king_position)) {
            return true;
        }
    }
    return false;
}

void ChessBoard::update_game_state() {
    Bitboard all_moves = Bitboard(0);
    for (auto from : our_pieces()) {
        all_moves |= generate_legal_moves(from);
    }
    if (all_moves.empty()) {
        if (is_player_in_check(player_)) {
            // Checkmate
            if (player_ == Player::White) {
                game_state_ = GameState::BlackWin;
            } else if (player_ == Player::Black) {
                game_state_ = GameState::WhiteWin;
            }
        } else {
            // Stalemate
            game_state_ = GameState::Draw;
        }
    }
    // Insufficient material
    if (!has_mating_material()) {
        game_state_ = GameState::Draw;
    }
    // 50-move rule
    if (fifty_move_rule_ == 100) {
        game_state_ = GameState::Draw;
    }
    // check for three-fold repetition
    if (get_repetition_count() >= 2) {
        game_state_ = GameState::Draw;
    }
}

bool ChessBoard::has_mating_material() const {
    if (!pawns_.empty() || !rooks_.empty() || !queens_.empty()) {
        return true;
    }

    int num_white_bishops = (bishops_ & white_pieces_).count();
    int num_black_bishops = (bishops_ & black_pieces_).count();
    int num_white_knights = (knights_ & white_pieces_).count();
    int num_black_knights = (knights_ & black_pieces_).count();

    if (num_white_bishops + num_white_knights > 1 ||
        num_black_bishops + num_black_knights > 1) {
        return true;
    }

    return false;
}

void ChessBoard::update_draw_condition(Square from, Square to) {
    fifty_move_rule_++;
    // reset if pawn is moved or piece is captured
    if (pawns_.get(from) || all_pieces_.get(to)) {
        fifty_move_rule_ = 0;
    }
}

void ChessBoard::castling(Square from, Square to) {
    // remove castling rights if king or rook is moved or captured
    if (from == 0 || to == 0) {
        castling_rights_ &= ~2;
    } else if (from == 7 || to == 7) {
        castling_rights_ &= ~1;
    } else if (from == 4 || to == 4) {
        castling_rights_ &= ~3;
    } else if (from == 56 || to == 56) {
        castling_rights_ &= ~8;
    } else if (from == 60 || to == 60) {
        castling_rights_ &= ~12;
    } else if (from == 63 || to == 63) {
        castling_rights_ &= ~4;
    }

    // move rook if castling
    if (kings_.get(from) && abs(from.file_ - to.file_) == 2) {
        if (from == 4) {
            if (to == 2) {
                rooks_.update(0, 3);
                white_pieces_.update(0, 3);
                all_pieces_.update(0, 3);
            } else if (to == 6) {
                rooks_.update(7, 5);
                white_pieces_.update(7, 5);
                all_pieces_.update(7, 5);
            }
        } else if (from == 60) {
            if (to == 58) {
                rooks_.update(56, 59);
                black_pieces_.update(56, 59);
                all_pieces_.update(56, 59);
            } else if (to == 62) {
                rooks_.update(63, 61);
                black_pieces_.update(63, 61);
                all_pieces_.update(63, 61);
            }
        }
    }
}

void ChessBoard::set_fen(std::string fen) {
    std::string board;
    std::istringstream fen_str(fen);
    fen_str >> board;

    int rank = 7;
    int file = 0;
    for (char c : board) {
        int i = rank * 8 + file;
        if (c == '/') {
            rank--;
            file = 0;
            continue;
        }
        if (std::isdigit(c)) {
            file += c - '0';
            continue;
        }
        if (std::isupper(c)) {
            white_pieces_.set(i);
        } else {
            black_pieces_.set(i);
        }
        all_pieces_.set(i);
        if (c == 'k' || c == 'K') {
            kings_.set(i);
        } else if (c == 'q' || c == 'Q') {
            queens_.set(i);
        } else if (c == 'r' || c == 'R') {
            rooks_.set(i);
        } else if (c == 'b' || c == 'B') {
            bishops_.set(i);
        } else if (c == 'n' || c == 'N') {
            knights_.set(i);
        } else if (c == 'p' || c == 'P') {
            pawns_.set(i);
        }
        file++;
    }

    // set player
    std::string player_string = "w";
    if (!fen_str.eof())
        fen_str >> player_string;
    player_ = player_string == "w" ? Player::White : Player::Black;

    // set castling rights
    std::string castling_rights_string = "-";
    if (!fen_str.eof())
        fen_str >> castling_rights_string;
    castling_rights_ = 0;
    for (char c : castling_rights_string) {
        if (c == 'K') {
            castling_rights_ |= 1;
        } else if (c == 'Q') {
            castling_rights_ |= 2;
        } else if (c == 'k') {
            castling_rights_ |= 4;
        } else if (c == 'q') {
            castling_rights_ |= 8;
        }
    }

    // set en passant square
    std::string en_passant_string = "-";
    if (!fen_str.eof())
        fen_str >> en_passant_string;
    if (en_passant_string != "-") {
        en_passant_.set(Square(en_passant_string));
    }

    if (!fen_str.eof())
        fen_str >> fifty_move_rule_;
    if (!fen_str.eof())
        fen_str >> fullmove_number_;
}

uint64_t ChessBoard::generate_hash() const {
    uint64_t hash = 0;
    for (auto i : white_pieces_) {
        if (pawns_.get(i)) {
            hash ^= PieceKeys[0][0][i.square_];
        } else if (knights_.get(i)) {
            hash ^= PieceKeys[0][1][i.square_];
        } else if (bishops_.get(i)) {
            hash ^= PieceKeys[0][2][i.square_];
        } else if (rooks_.get(i)) {
            hash ^= PieceKeys[0][3][i.square_];
        } else if (queens_.get(i)) {
            hash ^= PieceKeys[0][4][i.square_];
        } else if (kings_.get(i)) {
            hash ^= PieceKeys[0][5][i.square_];
        }
    }
    for (auto i : black_pieces_) {
        if (pawns_.get(i)) {
            hash ^= PieceKeys[1][0][i.square_];
        } else if (knights_.get(i)) {
            hash ^= PieceKeys[1][1][i.square_];
        } else if (bishops_.get(i)) {
            hash ^= PieceKeys[1][2][i.square_];
        } else if (rooks_.get(i)) {
            hash ^= PieceKeys[1][3][i.square_];
        } else if (queens_.get(i)) {
            hash ^= PieceKeys[1][4][i.square_];
        } else if (kings_.get(i)) {
            hash ^= PieceKeys[1][5][i.square_];
        }
    }
    if (!en_passant_.empty()) {
        int en_passant_square = en_passant_.getLSB();
        hash ^= EnPassantKeys[en_passant_square % 8];
    }
    if (castling_rights_ & 1) {
        hash ^= CastleKeys[0];
    }
    if (castling_rights_ & 2) {
        hash ^= CastleKeys[1];
    }
    if (castling_rights_ & 4) {
        hash ^= CastleKeys[2];
    }
    if (castling_rights_ & 8) {
        hash ^= CastleKeys[3];
    }
    if (player_ == Player::White) {
        hash ^= whiteToMoveKey;
    }

    return hash;
}

std::vector<uint64_t> ChessBoard::get_position_info() const {
    std::vector<uint64_t> position;
    position.push_back((white_pieces_ & pawns_).bitboard_);
    position.push_back((black_pieces_ & pawns_).bitboard_);
    position.push_back((white_pieces_ & bishops_).bitboard_);
    position.push_back((black_pieces_ & bishops_).bitboard_);
    position.push_back((white_pieces_ & rooks_).bitboard_);
    position.push_back((black_pieces_ & rooks_).bitboard_);
    position.push_back((white_pieces_ & queens_).bitboard_);
    position.push_back((black_pieces_ & queens_).bitboard_);
    position.push_back((white_pieces_ & kings_).bitboard_);
    position.push_back((black_pieces_ & kings_).bitboard_);
    position.push_back((white_pieces_ & knights_).bitboard_);
    position.push_back((black_pieces_ & knights_).bitboard_);

    int repetitions = get_repetition_count();
    if (repetitions == 0) {
        position.push_back(0);
        position.push_back(0);
    } else if (repetitions == 1) {
        position.push_back(1);
        position.push_back(0);
    } else {
        position.push_back(1);
        position.push_back(1);
    }

    return position;
}

File: engine/src/chessboard.rs
use std::fmt;

use crate::bitboard::Bitboard;
use crate::game::Player;
use crate::r#move::Move;
use crate::pieces::{Piece, Pieces};
use crate::square::Square;

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct StateInfo {
    pub captured_piece: Option<(Piece, bool)>,
    pub prev_en_passant: Bitboard,
    pub prev_castling_rights: u8,
    pub prev_fifty_move_rule: u8,
}

#[derive(Default, Debug, Clone, PartialEq)]
pub struct ChessBoard {
    fifty_move_rule: u8,
    castling_rights: u8,
    player: Player,
    pieces: Pieces,
    state_info: StateInfo,
}

impl ChessBoard {
    pub fn new() -> Self {
        let pieces = Pieces::new();
        let state_info = StateInfo::default();
        let player = Player::White;

        Self {
            fifty_move_rule: 0,
            castling_rights: 0b1111,
            player,
            pieces,
            state_info,
        }
    }

    pub fn load_from_fen(fen: String) -> Self {
        let mut chessboard = Self::default();
        let mut parts = fen.split_whitespace();

        let board = parts.next().unwrap_or("");
        let mut rank = 7;
        let mut file = 0;

        for c in board.chars() {
            let i: u8 = rank * 8 + file;

            if c == '/' {
                rank -= 1;
                file = 0;
                continue;
            }

            if c.is_ascii_digit() {
                file += c.to_digit(10).unwrap() as u8;
                continue;
            }

            if let Some((piece, is_white)) = Piece::from_char(c) {
                chessboard.pieces.set(piece, is_white, i);
            }

            file += 1;
        }

        let player_str = parts.next().unwrap_or("w");
        if player_str == "b" {
            chessboard.player = Player::Black;
        }

        let castling_str = parts.next().unwrap_or("-");
        chessboard.castling_rights = 0u8;
        for c in castling_str.chars() {
            match c {
                'K' => chessboard.castling_rights |= 1,
                'Q' => chessboard.castling_rights |= 2,
                'k' => chessboard.castling_rights |= 4,
                'q' => chessboard.castling_rights |= 8,
                _ => {}
            }
        }

        let en_passant_str = parts.next().unwrap_or("-");
        if en_passant_str != "-" {
            let square = en_passant_str.parse::<Square>().unwrap();
            chessboard.pieces.en_passant.set_square(square);
        }

        chessboard.fifty_move_rule = parts.next().and_then(|s| s.parse::<u8>().ok()).unwrap_or(0);

        chessboard
    }

    pub fn act(&mut self, r#move: Move) {
        let from = r#move.from;
        let to = r#move.to;
        let promotion = r#move.promotion;

        self.state_info.captured_piece = self.pieces.get_piece(to.square);
        self.state_info.prev_en_passant = self.pieces.en_passant;
        self.state_info.prev_castling_rights = self.castling_rights;
        self.state_info.prev_fifty_move_rule = self.fifty_move_rule;

        self.fifty_move_rule += 1;
        if self.pieces.pawns.get_square(from) || self.pieces.all_pieces.get_square(to) {
            self.fifty_move_rule = 0;
        }

        self.pieces.promote(promotion, from);
        self.pieces.update_en_passant(from, to);
        self.castle(from, to);

        self.pieces.update(from, to);
        self.player = self.player.switch();
    }

    pub fn undo(&mut self, r#move: Move) {
        let from = r#move.from;
        let to = r#move.to;
        let promotion = r#move.promotion;
        if promotion.is_some() {
            self.pieces.undo_promote(to);
        }

        self.pieces.update(to, from);
        if let Some((piece, is_white)) = self.state_info.captured_piece {
            self.pieces.set(piece, is_white, to.square);
        }
        self.undo_castle(from, to);

        if self.state_info.prev_en_passant.get_square(to) && self.pieces.pawns.get_square(from) {
            let captured_square = Square::new(from.rank, to.file);
            self.pieces.set(
                Piece::Pawn,
                self.player == Player::White,
                captured_square.square,
            );
        }

        self.pieces.en_passant = self.state_info.prev_en_passant;
        self.fifty_move_rule = self.state_info.prev_fifty_move_rule;
        self.player = self.player.switch();
    }

    pub fn get_pieces(&self) -> Pieces {
        self.pieces
    }

    pub fn get_our_pieces(&self) -> Bitboard {
        if self.player == Player::White {
            self.pieces.white_pieces
        } else {
            self.pieces.black_pieces
        }
    }

    pub fn get_player(&self) -> Player {
        self.player
    }

    pub fn get_castling_rights(&self) -> u8 {
        self.castling_rights
    }

    pub fn castle(&mut self, from: Square, to: Square) {
        // remove castling rights if king or rook is moved or captured
        if from == 0 || to == 0 {
            self.castling_rights &= !2;
        } else if from == 7 || to == 7 {
            self.castling_rights &= !1;
        } else if from == 4 || to == 4 {
            self.castling_rights &= !3;
        } else if from == 56 || to == 56 {
            self.castling_rights &= !8;
        } else if from == 60 || to == 60 {
            self.castling_rights &= !12;
        } else if from == 63 || to == 63 {
            self.castling_rights &= !4;
        }
        // move rook if castling
        if self.pieces.kings.get_square(from) && (from.square as i8 - to.square as i8).abs() == 2 {
            if from == 4 {
                if to == 2 {
                    self.pieces.rooks.update(0, 3);
                    self.pieces.white_pieces.update(0, 3);
                    self.pieces.all_pieces.update(0, 3);
                } else if to == 6 {
                    self.pieces.rooks.update(7, 5);
                    self.pieces.white_pieces.update(7, 5);
                    self.pieces.all_pieces.update(7, 5);
                }
            } else if from == 60 {
                if to == 58 {
                    self.pieces.rooks.update(56, 59);
                    self.pieces.black_pieces.update(56, 59);
                    self.pieces.all_pieces.update(56, 59);
                } else if to == 62 {
                    self.pieces.rooks.update(63, 61);
                    self.pieces.black_pieces.update(63, 61);
                    self.pieces.all_pieces.update(63, 61);
                }
            }
        }
    }

    fn undo_castle(&mut self, from: Square, to: Square) {
        self.castling_rights = self.state_info.prev_castling_rights;
        if self.pieces.kings.get_square(from) && (from.square as i8 - to.square as i8).abs() == 2 {
            if from == 4 {
                if to == 2 {
                    self.pieces.rooks.update(3, 0);
                    self.pieces.white_pieces.update(3, 0);
                    self.pieces.all_pieces.update(3, 0);
                } else if to == 6 {
                    self.pieces.rooks.update(5, 7);
                    self.pieces.white_pieces.update(5, 7);
                    self.pieces.all_pieces.update(5, 7);
                }
            } else if from == 60 {
                if to == 58 {
                    self.pieces.rooks.update(59, 56);
                    self.pieces.black_pieces.update(59, 56);
                    self.pieces.all_pieces.update(59, 56);
                } else if to == 62 {
                    self.pieces.rooks.update(61, 63);
                    self.pieces.black_pieces.update(61, 63);
                    self.pieces.all_pieces.update(61, 63);
                }
            }
        }
    }

    pub fn is_draw(&self) -> bool {
        !self.pieces.has_mating_material()
            || self.fifty_move_rule == 100
            || self.get_repetition_count() >= 2
    }

    pub fn get_repetition_count(&self) -> u8 {
        0
    }
}

impl fmt::Display for ChessBoard {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut board = String::new();

        for i in 0..64 {
            let c = self.pieces.get_char(i);

            board.push(c);

            if i % 8 == 7 {
                board.push('\n');
            } else {
                board.push(' ');
            }
        }

        for rank in (0..8).rev() {
            write!(f, "{}  ", rank + 1)?;
            let start = rank * 16;
            let end = start + 16;
            write!(f, "{}", &board[start..end])?;
        }

        write!(f, "   a b c d e f g h\n\n")
    }
}

#[cfg(test)]
mod tests {
    use super::ChessBoard;

    #[test]
    fn test_chessboard_to_string() {
        let board: ChessBoard = ChessBoard::new();
        let board_string = format!("{}", board);

        let expected = r#"8  r n b q k b n r
7  p p p p p p p p
6  . . . . . . . .
5  . . . . . . . .
4  . . . . . . . .
3  . . . . . . . .
2  P P P P P P P P
1  R N B Q K B N R
   a b c d e f g h

"#;

        assert_eq!(board_string, expected);
    }

    #[test]
    fn test_load_from_fen() {
        let board_from_fen = ChessBoard::load_from_fen(
            "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1".to_string(),
        );
        let new_board = ChessBoard::new();

        assert_eq!(board_from_fen, new_board);
    }
}
